From ef84b868d164752fa1f45d392ec9fdc916a69d9c Mon Sep 17 00:00:00 2001
From: Prasad Boddupalli <pboddupalli@netlogicmicro.com>
Date: Fri, 13 Aug 2010 18:23:36 -0700
Subject: [PATCH 094/565] Change the xensp field

Currently, the xensp field in the shared page of domains is a scalar
field and is initialized with the address of the xensp[] entry
corresponding to the primary cpu. A scalar value is inadequate for
smp domains as hypercalls could be called on secondary cpus too.
The commit fixes the problem by making the xensp field of the
shared_info page to an array.

Based on Broadcom SDK 2.3.

Signed-off-by: Prasad Boddupalli <pboddupalli@netlogicmicro.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 arch/mips/include/asm/xen/hypercall.h | 3 ++-
 arch/mips/include/asm/xen/interface.h | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/arch/mips/include/asm/xen/hypercall.h b/arch/mips/include/asm/xen/hypercall.h
index bf87006..5663ed3 100644
--- a/arch/mips/include/asm/xen/hypercall.h
+++ b/arch/mips/include/asm/xen/hypercall.h
@@ -48,6 +48,7 @@
 #include <xen/interface/physdev.h>
 #include <xen/interface/platform.h>
 #include <asm/xen/hypervisor.h>
+#include <asm/netlogic/mips-exts.h>
 
 /*
  * The hypercall asms have to meet several constraints:
@@ -122,7 +123,7 @@ extern struct { unsigned long _entry; } hypercall_page[];
 #define switch_xen_linux_stacks() \
 ({ \
 	load_gp(__get_cpu_var(gpctx)); \
-	store_gp((*(HYPERVISOR_shared_info->arch.xensp)) & ~(XEN_PAGE_SIZE - 1)); \
+	store_gp((*(HYPERVISOR_shared_info->arch.xensp[hard_smp_processor_id()])) & ~(XEN_PAGE_SIZE - 1)); \
 })
 
 #define restore_linux_stack() store_gp(__get_cpu_var(gpctx))
diff --git a/arch/mips/include/asm/xen/interface.h b/arch/mips/include/asm/xen/interface.h
index b3ee38c..25c8e4f 100644
--- a/arch/mips/include/asm/xen/interface.h
+++ b/arch/mips/include/asm/xen/interface.h
@@ -31,6 +31,7 @@
 
 #define __XEN_INTERFACE_VERSION__ __XEN_LATEST_INTERFACE_VERSION__
 
+#include <linux/threads.h>
 #include <asm/sgidefs.h>
 
 #ifdef __XEN__
@@ -116,7 +117,7 @@ struct arch_shared_info {
     /* Frame containing list of mfns containing list of mfns containing p2m. */
     xen_pfn_t     pfn_to_mfn_frame_list_list;
     unsigned long nmi_reason;
-	unsigned long *xensp;
+	unsigned long *xensp[NR_CPUS];
 	unsigned int dom_page_shift;
 	unsigned int pgtable_levels;
     uint64_t pad[32];
-- 
1.8.4.93.g57e4c17

